### ISSUE TITLE
FilePath should not cache stat() return value

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3160]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#3160|
|Type|defect|
|Created|2008-04-03 17:17:13Z|

Stat caching causes many correctness problems.

Perhaps we could add a mode or a subclass that turns it on, but the default should definitely be no caching.

<details><summary>Searchable metadata</summary>

```
trac-id__3160 3160
type__defect defect
reporter__radix radix
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__core core
keywords__ 
time__1207243033000000 1207243033000000
changetime__1207257657000000 1207257657000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

No.

This was a huge improvement in at least 2 different major applications of filepath.  Any outright removal of stat caching will simply reduce the likelihood of certain race conditions rather than make the system more correct.

If you want to correct the problems, describe the problems, and we wil fix then; I believe another ticket was filed about an interaction with mkdir() or something.  If we absolutely need to remove stat caching to fix an actual problem, then we can talk about some other performance improvement.  However, having destructive filesystem operations force a re-stat should fix almost every real issue.

Another way to think about this: a FilePath _IS_ a stat result.  Maybe it should timestamp the time it calls stat(); maybe there should be a more convenient way of getting a new stat result.  But regardless, we are not going to remove stat caching just for the sake of removing it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `closed`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

If I'm going to be arbitrary about this, it probably behooves me to explain my motivations a bit more clearly:

I don't have benchmarks, but the specific thing it optimizes is stuff like this:

```
#!py
f = FilePath("...").child(".some-interesting-cache")
if f.exists(): # one stat
    if f.isdir(): # two stat
        if f.getModificationTime() > someTime: # three stat
            ...
    elif f.getsize() > someSize:  # four
# five stat, six stat, seven stat, more...
```

This occurs, for example, scrounging around for dropin.cache files and other sundry places in plugin discovery and twisted.python.modules inspection.  It can drop the number of stats as much as 8 to 1.

This isn't an uncommon idiom, and at the very least the idiom:

```
#!py
for f in something.children():
    if f.isdir():
        something()
    if f.isfile(): # a bit redundant, but reads nicely
        somethingElse()
```

is pretty common in average filepath code, and stat caching halves the number of stats issued here; given the relative expense of the syscall it can as much as double the speed of the code.

All the stuff with internal inconsistency and FilePath is simply a bug, which should be fixed.  I am not defending the fact that mkdir(), isdir() on the same path object will be wrong; there is no efficiency gain there.  Everything destructive should be fixed to be internally consistent.

However, the extra long lifetime of FilePath race conditions have spotted a few concurrent-directory-modification bugs for me which normally would only have occurred under really high load in situations that would have been impossible to debug.  There isn't a long time between the two stats in .exists() and .isdir() usually, but with stat caching you get a result which is sort of ... consistently inconsistent, a stat result from a single fixed point in time, which is more obviously wrong than "why did they disagree this one time out of a million".  I actually _like_ this quirk, except for the fact that it's a surprise because I have to remember that stat results are being cached, and the slightly strange times at which they're cached, interface-wise.  Also, any lack of documentation of the semantics of stat caching is a bug.

So basically; any issue which stat caching causes, we can and should address.  Itamar had some interesting suggestions on ways to evolve the API that radix just mentioned to me, but I'm not going to fill in those here because this is already getting pretty long, but they might preserve the performance and repeatability stuff I like without creating as much confusion.  I closed this ticket as invalid because stat caching _itself_ is not the issue.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
